staging/src/k8s.io/apiserver/pkg/storage/etcd3/etcdfeature/feature_support_checker.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcdfeature
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/klog/v2"
+)
+
+var (
+	// Define these static versions to use for checking version of etcd, issue on kubernetes #123192
+	v3_4_25 = version.MustParseSemantic("3.4.25")
+	v3_5_0  = version.MustParseSemantic("3.5.0")
+	v3_5_8  = version.MustParseSemantic("3.5.8")
+
+	// DefaultFeatureSupportChecker is a shared global FeatureSupportChecker.
+	DefaultFeatureSupportChecker FeatureSupportChecker = newDefaultFeatureSupportChecker()
+)
+
+// FeatureSupportChecker to define Supports functions.
+type FeatureSupportChecker interface {
+	Supports(feature storage.Feature) (bool, error)
+	// This method works with etcd client to recalcuate feature support and caches it.
+	ClientSupports(ctx context.Context, c client, feature storage.Feature) (bool, error)
+	// Only use for testing and use defer afterwork.
+	SetFeatureSupportCheckerDuringTest(feature storage.Feature, value bool) func()
```

Comment:
I'm not a fan of it - I don't want to have test-specific methods for interfaces running in production.

Instead, we can have an extended implementation purely for tests (testWatchCache is an example of that pattern).

Reply:
I'll change this thanks for pointing this out.

Reply:
It seems tests will pass without this part, i remove it for now but i can re-implement it for tests if we need it in future.